sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,475 @@
+""" Kernels for Gaussian process regression and classification.
+
+The kernels in this module allow kernel-engineering, i.e., they can be
+combined via the "+" and "*" operators. These expressions can also contain
+scalar values, which are automatically converted to a constant kernel.
+
+All kernel allow (analytic) gradient-based hyperparameter optimization.
+The space of hyperparameters can be specified by giving lower und upper
+boundaries for the value of each hyperparameter (the search space is thus
+rectangular). This can be achieved by using a pair or triple instead of a
+single float wherever a parameter value is specified. In case of a pair,
+the first value specifies the lower boundary and the second value the upper
+boundary. In case of a triple, the middle value specified the initial value
+of the parameter during hyperparameter-optimization.
+"""
+
+# Author: Jan Hendrik Metzen <jhm@informatik.uni-bremen.de>
+# Licence: BSD 3 clause
+
+# Note: this module is strongly inspired by the kernel module of the george
+#       package.
+
+import numpy as np
+from scipy.spatial.distance import pdist, cdist, squareform
+
+
+class Kernel(object):
+    """ Base class for all kernels."""
+
+    def _parse_param_space(self, param_space):
+        if not hasattr(param_space, "__iter__"):  # fixed hyperparameter
+            self.params = np.array([float(param_space)])
+            self.has_bounds = False
+            return
+        param_space = np.atleast_2d(param_space)
+        if param_space.shape[1] == 1:  # fixed hyperparameter
+            self.params = param_space[:, 0]
+            self.has_bounds = False
+        elif param_space.shape[1] == 2:  # lower+upper bound for hyperparameter
+            self.bounds = param_space
+            self.has_bounds = True
+            # Use geometric mean of upper and lower boundary as initial
+            # hyperparameter value
+            assert not np.any(self.l_bound == None)  # XXX: enforce element-wise comparison to None
+            assert not np.any(self.u_bound == None)
+            self.params = np.array([np.sqrt(self.l_bound * self.u_bound)])
+        elif param_space.shape[1] == 3:  # lower bound, initial value, upper bound
+            self.params = param_space[:, 1]
+            self.bounds = param_space[:, [0, 2]]
+            self.has_bounds = True
+        else:
+            raise Exception()
```

Comment:
Put a description. It should probably be a `ValueError`. But this tuple business is tricky ...